### PR TITLE
Add array range support

### LINKED ISSUE #178: Range operator doesn't work with arrays
#### Describe the bug

The range operator works with strings and spans but not with arrays. The error is
> CS0656	Missing compiler required member 'System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray'

#### Minimal Repro

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>netstandard2.0</TargetFrameworks>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Polyfill" Version="5.1.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="System.Memory" Version="4.5.5" />
  </ItemGroup>

</Project>
```

```c#
public void SliceArray()
{
    var array = new[] { "aa", "bb", "cb" };

    // string compiles
    _ = array.First()[..2];

    // span compiles
    _ = array.AsSpan()[..2];

    // array doesn't compile
    _ = array[..2];
}
```

#### Workaround

Adding the [RuntimeHelpers.GetSubArray](https://github.com/dotnet/runtime/blob/v8.0.4/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L21C1-L49C10) solves the problem. Note that the behavior of this method changes in [.NET 9](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/9.0/getsubarray-return).

### PR DESCRIPTION
**Pull request by gao-artur**
Fixes #178 

@SimonCropp I tried adding `RuntimeHelpers.GetSubArray` (the pre .NET 9 version) to support array range, but for some reason, I'm getting a compiler error
image

### CODE REVIEW

**Comment by SimonCropp** [comment]
looks like a bug in roslyn. can you try rains a bug in https://github.com/dotnet/roslyn

**Comment by gao-artur** [comment]
Honestly, I have no idea how to describe this bug. I'm not sure they will love the reference to this PR as a repro...

**Comment by gao-artur** [comment]
This is the exception, and it's thrown by all the projects with `<Import Project="$(SolutionDir)\TestIncludes.targets" />`

> System.NullReferenceException: Object reference not set to an instance of an object.
>    at Microsoft.CodeAnalysis.CodeGen.ReferenceDependencyWalker.VisitSignature(ISignature signature, EmitContext context)
>    at Microsoft.CodeAnalysis.Emit.CommonPEModuleBuilder.GetFakeSymbolTokenForIL(ISignature symbol, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
>    at Microsoft.CodeAnalysis.CodeGen.ILBuilder.EmitArrayBlockInitializer(ImmutableArray`1 data, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.<TryEmitOptimizedReadonlySpanCreation>g__tryEmitAsCachedArrayFromBlob|88_1(NamedTypeSymbol spanType, BoundExpression wrappedExpression, Int32 elementCount, ImmutableArray`1 data, ArrayTypeSymbol& arrayType, TypeSymbol elementType)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.TryEmitOptimizedReadonlySpanCreation(NamedTypeSymbol spanType, BoundExpression wrappedExpression, Boolean used, BoundExpression inPlaceTarget, Boolean& avoidInPlace, BoundExpression start, BoundExpression length)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.TryEmitOptimizedReadonlySpan(BoundObjectCreationExpression expression, Boolean used, BoundExpression inPlaceTarget, Boolean& avoidInPlace)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitObjectCreationExpression(BoundObjectCreationExpression expression, Boolean used)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArguments(ImmutableArray`1 arguments, ImmutableArray`1 parameters, ImmutableArray`1 argRefKindsOpt)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStaticCallExpression(BoundCall call, UseKind useKind)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitSequenceExpression(BoundSequence sequence, Boolean used)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCoreWithStackGuard(BoundExpression expression, Boolean used)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatementAndCountInstructions(BoundStatement statement)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitSequencePointStatement(BoundSequencePoint node)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatements(ImmutableArray`1 statements)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitUninstrumentedBlock(BoundBlock block)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatements(ImmutableArray`1 statements)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitUninstrumentedBlock(BoundBlock block)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatementList(BoundStatementList list)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitStatement(BoundStatement statement)
>    at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GenerateImpl()
>    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.GenerateMethodBody(PEModuleBuilder moduleBuilder, MethodSymbol method, Int32 methodOrdinal, BoundStatement block, ImmutableArray`1 lambdaDebugInfo, ImmutableArray`1 orderedLambdaRuntimeRudeEdits, ImmutableArray`1 closureDebugInfo, ImmutableArray`1 stateMachineStateDebugInfos, StateMachineTypeSymbol stateMachineTypeOpt, VariableSlotAllocator variableSlotAllocatorOpt, BindingDiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, ImportChain importChainOpt, Boolean emittingPdb, ImmutableArray`1 codeCoverageSpans, AsyncForwardEntryPoint entryPointOpt)
>    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
>    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
>    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass25_0.<CompileNamedTypeAsync>b__0()

**Comment by SimonCropp** [comment]
> Honestly, I have no idea how to describe this bug. I'm not sure they will love the reference to this PR as a repro...

then here is a minimal repro https://github.com/SimonCropp/RuntimeHelpersRepro

File: src/Polyfill/RuntimeHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+static partial class RuntimeHelpers
+{
+    /// <summary>
+    /// Provides a fast way to initialize an array from data that is stored in a module.
+    /// </summary>
+    /// <param name="array">The array to be initialized.</param>
+    /// <param name="fldHandle">A field handle that specifies the location of the data used to initialize the array</param>
+    [Link("https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.initializearray")]
+    [Intrinsic]
+    public static extern void InitializeArray(Array array, RuntimeFieldHandle fldHandle);
```
Comment (gao-artur) [inline src/Polyfill/RuntimeHelpers.cs:29]: @SimonCropp, after adding this method, everything compiles and tests are green. But I have a few questions:

1. This method should already exist in all the TFMs. Can it cause runtime errors?
image

2. While all tests run fine, the ReSharper runner shows the following errors. Any idea how to fix them? Is it safe to ignore?
image

3. I created 2 partial `RuntimeHelpers` classes, as `InitializeArray` is not necessarily related to the `IndexRange` feature and theoretically can be required by other features added in the future. Is it ok, or do you prefer a single class?

File: src/Polyfill/IndexRange/RuntimeHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        if (length == 0)
+        {
+            return [];
+        }
+
+        var dest = new T[length];
+
+        Array.Copy(array, offset, dest, 0, length);
+
+        return dest;
```
Comment (gao-artur) [inline src/Polyfill/IndexRange/RuntimeHelpers.cs:44]: Do you prefer this implementation or one that will be released with .NET 9?
https://github.com/dotnet/runtime/blob/v9.0.0-preview.3.24172.9/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L47-L54

File: src/Polyfill/RuntimeHelpers.cs (hunk 1 above)
Comment (SimonCropp) [reply src/Polyfill/RuntimeHelpers.cs:29]: try this
```
// <auto-generated />

#if (NET46X && FeatureValueTuple) || NET47X || NET48X || NETSTANDARD2_0 || NETCOREAPP2X

namespace System.Runtime.CompilerServices;

using System.Reflection;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using Link = System.ComponentModel.DescriptionAttribute;

/// <summary>
/// Provides a set of static methods and properties that provide support for compilers.
/// This class cannot be inherited.
/// </summary>
[ExcludeFromCodeCoverage]
[DebuggerNonUserCode]
#if PolyPublic
public
#endif
static partial class RuntimeHelpers
{
    static readonly Action<Array,RuntimeFieldHandle> initializeArrayAction;

    static RuntimeHelpers()
    {
        //CompilationRelaxations is co located with RuntimeHelpers
        var runtimeHelpersType = typeof(CompilationRelaxations)
            .Assembly
            .GetType("System.Runtime.CompilerServices.RuntimeHelpers");
        var initializeArrayMethod = runtimeHelpersType.GetMethod(
            "InitializeArray",
            BindingFlags.Static | BindingFlags.Public,
            null,
            [typeof(Array), typeof(RuntimeFieldHandle)],
            null)!;
        initializeArrayAction = (Action<Array,RuntimeFieldHandle>)initializeArrayMethod.CreateDelegate(typeof(Action<Array,RuntimeFieldHandle>));
    }

    /// <summary>
    /// Provides a fast way to initialize an array from data that is stored in a module.
    /// </summary>
    /// <param name="array">The array to be initialized.</param>
    /// <param name="fldHandle">A field handle that specifies the location of the data used to initialize the array</param>
    [Link("https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.initializearray")]
    public static void InitializeArray(Array array, RuntimeFieldHandle fldHandle) =>
        initializeArrayAction(array, fldHandle);
}
#endif
```

and then we dont need IntrinsicAttribute

**Comment by SimonCropp** [comment]
ok. going to merge this one. i will clean up the last remaining things after. thanks for all the research

**Pull request merged**